Pallet liquidity rewards impl

lemunozm opened the pull request:
# Description

Implements [`liquidity-pallet-rewards`](https://centrifuge.hackmd.io/9T9zAZaTRD6ZcHN-Gd_lzw?view#pallet-liquidity-rewards-spec)

Fixes #999

## Changes and Descriptions

- [x] Implement extrinsics
- [x] on_initialized hook
- [x] Weights
- [x] Mock
- [x] Tests
- [x] Events
- [x] Docs
- [x] Readme doc entry
- [x] Benchmarks

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# How Has This Been Tested?

```sh
cargo test -p pallet-liquidity-rewards --features runtime-benchmarks
PARA_CHAIN_SPEC=development ./scripts/init.sh benchmark pallet-liquidity-rewards
```

# Checklist:

- [x] I have added Rust doc comments to structs, enums, traits and functions
- [x] I have made corresponding changes to the documentation
- [x] I have performed a self-review of my code
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [ ] I rebased on the latest `main` branch

mikiquantum reviewed (commented):
Overall is good and simple, I will have another look as soon as the other tasks listed in the description are in 👍

lemunozm commented on pallets/liquidity-rewards/src/lib.rs:
```diff
@@ -0,0 +1,263 @@
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	pub(super) type GroupWeights<T: Config> =
```
The `Groups` storage of `pallet-rewards` stores the `Group` data for each `GroupId` (without weight information).

This new storage does two things:
  - Stores the weight of each group
  - Identify the `GroupId` used only by this pallet. The length of this storage can be (and will be) different to the length of `Groups` in `pallet-rewards` where all groups used in other pallets will be stored there.

mikiquantum replied:
ok, I see, I saw the T::Rate type in the pallet rewards, and thought that was the weight, but is the rpt. It makes sense that the rewards pallet doesnt need to know about an implementation that uses weights for distribution.

lemunozm commented on pallets/liquidity-rewards/src/lib.rs:
```diff
@@ -0,0 +1,263 @@
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
```
Totally, the only "complicated" place is `on_initialize()`.

lemunozm commented:
@Miguel changes from your last review:
- Added a mock for rewards traits using [`mockall`](https://docs.rs/mockall/latest/mockall/index.html) (a cool mocking library).
- Simplified `mock.rs` a lot (thanks to mocking the reward trait).
- Added tests/benchmarks/docs.
- Modified some storage to improve the performance at `on_initialize()`
  - `GroupWeights` storage is now a simple `BoundedBTreeMap` in a `StorageValue`.
  - `EpochChanges` is a new struct that combines all different epoch changes and is based on `BoundedBTreeMap` instead of using several `StorageMap`.
  - `ends_on` attribute of epoch extract to its own struct to reduce the deserialization phase each `on_initialize`. There is no need to deserialize the whole epoch if 99% of `on_initialize()` calls will not use the rest of the `Epoch` data.
- Added `pallet-liquidity-pallet` to `development` runtime (needed to generate weights)

mikiquantum commented on pallets/liquidity-rewards/src/lib.rs:
```diff
@@ -0,0 +1,412 @@
+			group_id: T::GroupId,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+			NextEpochChanges::<T>::try_mutate(|changes| {
```
nit: spacing

mikiquantum commented on pallets/liquidity-rewards/src/lib.rs:
```diff
@@ -0,0 +1,412 @@
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	pub(super) type EndOfEpoch<T: Config> = StorageValue<
```
Maybe we can add a comment explaining that we took the endOfEpoch out of the active epoch data storage to not have to deserialize the whole struct for 99% of the time assuming a duration of 100 blocks :)

lemunozm replied:
I like it!

mikiquantum reviewed (approved):
LGTM!

mustermeiszer commented on runtime/development/Cargo.toml:
```diff
@@ -287,6 +292,7 @@ try-runtime = [
     "pallet-fees/try-runtime",
     "pallet-identity/try-runtime",
     "pallet-interest-accrual/try-runtime",
+    "pallet-liquidity-rewards/runtime-benchmarks",
```
We must be a bit more cautious with feature flags @lemunozm and @mikiquantum. I guess development would not build with try-runtime in this scenario separately. Just am marking this here, as these feature flags cost us quite some time, then they are set wrongly and some way down the line in the future make problems. 

This also goes for myself. Maybe we should brainstorm, how we could enforce correct feature flags and what should be the default setup for pallets and their features.
